File: backend/modules/user/src/main/java/com/micoach/user/domain/UserProfile.java
package com.micoach.user.domain;

import lombok.Getter;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.Instant;
import java.util.List;

/**
 * Perfil de salud del usuario (tabla user_profiles). 1:1 con auth_users.
 * Valores aquí son iniciales/actuales; el histórico vive en progress_entries.
 */
@Getter
public class UserProfile {

    private Long id;
    private final Long userId;
    private String sex;
    private LocalDate birthDate;
    private BigDecimal heightCm;
    private BigDecimal weightKg;
    private String activityLevel;
    private String experienceLevel;
    private List<String> equipment;
    private Integer trainingDaysPerWeek;
    private Integer trainingMinutes;
    private String preferredTime;
    private String timezone;
    private Integer tdeeCalories;
    private String dietaryGoal;
    private String notes;
    private final Instant createdAt;
    private Instant updatedAt;

    private UserProfile(Long id, Long userId, String sex, LocalDate birthDate,
                        BigDecimal heightCm, BigDecimal weightKg, String activityLevel,
                        String experienceLevel, List<String> equipment, Integer trainingDaysPerWeek,
                        Integer trainingMinutes, String preferredTime, String timezone,
                        Integer tdeeCalories, String dietaryGoal, String notes,
                        Instant createdAt, Instant updatedAt) {
        this.id = id;
        this.userId = userId;
        this.sex = sex;
        this.birthDate = birthDate;
        this.heightCm = heightCm;
        this.weightKg = weightKg;
        this.activityLevel = activityLevel;
        this.experienceLevel = experienceLevel;
        this.equipment = equipment;
        this.trainingDaysPerWeek = trainingDaysPerWeek;
        this.trainingMinutes = trainingMinutes;
        this.preferredTime = preferredTime;
        this.timezone = timezone;
        this.tdeeCalories = tdeeCalories;
        this.dietaryGoal = dietaryGoal;
        this.notes = notes;
        this.createdAt = createdAt;
        this.updatedAt = updatedAt;
    }

    public static UserProfile empty(Long userId) {
        Instant now = Instant.now();
        return new UserProfile(null, userId, null, null, null, null, null, null,
                List.of(), null, null, null, null, null, null, null, now, now);
    }

    public static UserProfile restore(Long id, Long userId, String sex, LocalDate birthDate,
                                      BigDecimal heightCm, BigDecimal weightKg, String activityLevel,
                                      String experienceLevel, List<String> equipment,
                                      Integer trainingDaysPerWeek, Integer trainingMinutes,
                                      String preferredTime, String timezone, Integer tdeeCalories,
                                      String dietaryGoal, String notes,
                                      Instant createdAt, Instant updatedAt) {
        return new UserProfile(id, userId, sex, birthDate, heightCm, weightKg, activityLevel,
                experienceLevel, equipment == null ? List.of() : equipment, trainingDaysPerWeek,
                trainingMinutes, preferredTime, timezone, tdeeCalories, dietaryGoal, notes,
                createdAt, updatedAt);
    }

    public void update(String sex, LocalDate birthDate, BigDecimal heightCm, BigDecimal weightKg,
                       String activityLevel, String experienceLevel, List<String> equipment,
                       Integer trainingDaysPerWeek, Integer trainingMinutes, String preferredTime,
                       String timezone, Integer tdeeCalories, String dietaryGoal, String notes) {
        this.sex = sex;
        this.birthDate = birthDate;
        this.heightCm = heightCm;
        this.weightKg = weightKg;
        this.activityLevel = activityLevel;
        this.experienceLevel = experienceLevel;
        this.equipment = equipment;
        this.trainingDaysPerWeek = trainingDaysPerWeek;
        this.trainingMinutes = trainingMinutes;
        this.preferredTime = preferredTime;
        this.timezone = timezone;
        this.tdeeCalories = tdeeCalories;
        this.dietaryGoal = dietaryGoal;
        this.notes = notes;
        this.updatedAt = Instant.now();
    }
}

File: mobile/lib/features/profile/domain/user_profile.dart
// =====================================================================
// MiCoach — Perfil de salud (mirror de UserDtos.ProfileResponse en backend).
// =====================================================================
class UserProfile {
  final int? id;
  final String? sex;
  final DateTime? birthDate;
  final double? heightCm;
  final double? weightKg;
  final String? activityLevel;
  final String? experienceLevel;
  final List<String> equipment;
  final int? trainingDaysPerWeek;
  final int? trainingMinutes;
  final String? preferredTime;
  final String? timezone;
  final int? tdeeCalories;
  final String? dietaryGoal;
  final String? notes;

  const UserProfile({
    this.id,
    this.sex,
    this.birthDate,
    this.heightCm,
    this.weightKg,
    this.activityLevel,
    this.experienceLevel,
    this.equipment = const [],
    this.trainingDaysPerWeek,
    this.trainingMinutes,
    this.preferredTime,
    this.timezone,
    this.tdeeCalories,
    this.dietaryGoal,
    this.notes,
  });

  factory UserProfile.fromJson(Map<String, dynamic> json) {
    return UserProfile(
      id: json['id'] as int?,
      sex: json['sex'] as String?,
      birthDate: json['birthDate'] != null ? DateTime.parse(json['birthDate'] as String) : null,
      heightCm: (json['heightCm'] as num?)?.toDouble(),
      weightKg: (json['weightKg'] as num?)?.toDouble(),
      activityLevel: json['activityLevel'] as String?,
      experienceLevel: json['experienceLevel'] as String?,
      equipment: (json['equipment'] as List<dynamic>? ?? const []).map((e) => e as String).toList(),
      trainingDaysPerWeek: json['trainingDaysPerWeek'] as int?,
      trainingMinutes: json['trainingMinutes'] as int?,
      preferredTime: json['preferredTime'] as String?,
      timezone: json['timezone'] as String?,
      tdeeCalories: json['tdeeCalories'] as int?,
      dietaryGoal: json['dietaryGoal'] as String?,
      notes: json['notes'] as String?,
    );
  }

  Map<String, dynamic> toJson() => {
        'sex': sex,
        'birthDate': birthDate?.toIso8601String().split('T').first,
        'heightCm': heightCm,
        'weightKg': weightKg,
        'activityLevel': activityLevel,
        'experienceLevel': experienceLevel,
        'equipment': equipment,
        'trainingDaysPerWeek': trainingDaysPerWeek,
        'trainingMinutes': trainingMinutes,
        'preferredTime': preferredTime,
        'timezone': timezone,
        'tdeeCalories': tdeeCalories,
        'dietaryGoal': dietaryGoal,
        'notes': notes,
      };

  UserProfile copyWith({
    String? sex,
    DateTime? birthDate,
    double? heightCm,
    double? weightKg,
    String? activityLevel,
    String? experienceLevel,
    List<String>? equipment,
    int? trainingDaysPerWeek,
    int? trainingMinutes,
    String? preferredTime,
    String? timezone,
    int? tdeeCalories,
    String? dietaryGoal,
    String? notes,
  }) {
    return UserProfile(
      id: id,
      sex: sex ?? this.sex,
      birthDate: birthDate ?? this.birthDate,
      heightCm: heightCm ?? this.heightCm,
      weightKg: weightKg ?? this.weightKg,
      activityLevel: activityLevel ?? this.activityLevel,
      experienceLevel: experienceLevel ?? this.experienceLevel,
      equipment: equipment ?? this.equipment,
      trainingDaysPerWeek: trainingDaysPerWeek ?? this.trainingDaysPerWeek,
      trainingMinutes: trainingMinutes ?? this.trainingMinutes,
      preferredTime: preferredTime ?? this.preferredTime,
      timezone: timezone ?? this.timezone,
      tdeeCalories: tdeeCalories ?? this.tdeeCalories,
      dietaryGoal: dietaryGoal ?? this.dietaryGoal,
      notes: notes ?? this.notes,
    );
  }
}
